src/ol/reproj/tile.js

```diff
@@ -0,0 +1,340 @@
+      !ol.extent.intersects(maxSourceExtent, srcExtent)) {
+    this.state = ol.TileState.EMPTY;
+  } else {
+    var srcRange = sourceTileGrid.getTileRangeForExtentAndZ(
+        srcExtent, this.srcZ_);
+
+    var xRange = [];
+    var srcFullRange = sourceTileGrid.getFullTileRange(this.srcZ_);
+    if (srcFullRange) {
+      srcRange.minY = Math.max(srcRange.minY, srcFullRange.minY);
+      srcRange.maxY = Math.min(srcRange.maxY, srcFullRange.maxY);
+
+      if (srcRange.minX > srcRange.maxX) {
+        var i;
+        for (i = srcRange.minX; i <= srcFullRange.maxX; i++) {
+          xRange.push(i);
+        }
+        for (i = srcFullRange.minX; i <= srcRange.maxX; i++) {
+          xRange.push(i);
+        }
+      } else {
+        var first = Math.max(srcRange.minX, srcFullRange.minX);
+        var last = Math.min(srcRange.maxX, srcFullRange.maxX);
+        for (var j = first; j <= last; j++) {
+          xRange.push(j);
+        }
+      }
+    } else {
+      for (var k = srcRange.minX; k <= srcRange.maxX; k++) {
+        xRange.push(k);
+      }
+    }
+
+    var tilesRequired = xRange.length * srcRange.getHeight();
+    if (!goog.asserts.assert(tilesRequired < ol.RASTER_REPROJ_MAX_SOURCE_TILES,
+        'reasonable number of tiles is required')) {
+      this.state = ol.TileState.ERROR;
+      return;
+    }
+    xRange.forEach(function(srcX, i, arr) {
+      for (var srcY = srcRange.minY; srcY <= srcRange.maxY; srcY++) {
+        var tile = getTileFunction(this.srcZ_, srcX, srcY, pixelRatio);
+        if (tile) {
+          this.srcTiles_.push(tile);
+        }
+      }
+    }, this);
+
+    if (this.srcTiles_.length === 0) {
+      this.state = ol.TileState.EMPTY;
+    }
+  }
+};
+goog.inherits(ol.reproj.Tile, ol.Tile);
+
+
+/**
+ * @inheritDoc
+ */
+ol.reproj.Tile.prototype.disposeInternal = function() {
+  if (this.state == ol.TileState.LOADING) {
+    this.unlistenSources_();
+  }
+  goog.base(this, 'disposeInternal');
+};
+
+
+/**
+ * @inheritDoc
+ */
+ol.reproj.Tile.prototype.getImage = function(opt_context) {
+  if (opt_context !== undefined) {
+    var image;
+    var key = goog.getUid(opt_context);
+    if (key in this.canvasByContext_) {
+      return this.canvasByContext_[key];
+    } else if (goog.object.isEmpty(this.canvasByContext_)) {
+      image = this.canvas_;
+    } else {
+      image = /** @type {HTMLCanvasElement} */ (this.canvas_.cloneNode(false));
+    }
+    this.canvasByContext_[key] = image;
+    return image;
+  } else {
+    return this.canvas_;
+  }
+};
+
+
+/**
+ * @private
+ */
+ol.reproj.Tile.prototype.reproject_ = function() {
+  var sources = [];
+  this.srcTiles_.forEach(function(tile, i, arr) {
+    if (tile && tile.getState() == ol.TileState.LOADED) {
+      sources.push({
+        extent: this.sourceTileGrid_.getTileCoordExtent(tile.tileCoord),
+        image: tile.getImage()
+      });
+    }
+  }, this);
+
+  var tileCoord = this.getTileCoord();
```

Comment:
You might be able to remove most (if not all) of the wrapX complexity in `ol.reproj.render` if you use `this.getTileCoordForTileUrlFunction()` here instead of `this.getTileCoord()`.